client.go

```diff
@@ -0,0 +1,322 @@
+			wg.Done()
+		}()
+		triesSplit := atomic.LoadUint64(triesPtr)
+		coalesceErr2 := c.write(ctx, bucket, org, &triesSplit, m[(len(m)<<2):]...)
+		wg.Wait()
+		var cerr coalescingError
+		if coalesceErr1 != nil {
+			cerr = append(cerr, coalesceErr1)
+		}
+		if coalesceErr2 != nil {
+			err = append(cerr, coalesceErr2)
+		}
+		return cerr
+
+	default:
+		resp.Body.Close()
+		return &genericRespError{
+			Code:    resp.Status,
+			Message: "internal server error",
+		}
+
+	}
+	defer func() {
+		err2 := resp.Body.Close()
+		if err == nil && err2 != nil {
+			err = err2
+		}
+	}()
+	e.FailOnFieldErr(c.errOnFieldErr)
+
+	for i := range m {
+		if _, err = e.Encode(m[i]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// backoff is a helper method for backoff, triesPtr must not be nil.
+func (c *Client) backoff(triesPtr *uint64, resp *http.Response, err error) error {
+	tries := atomic.LoadUint64(triesPtr)
+	if c.maxRetries >= 0 || int(tries) >= c.maxRetries {
+		return maxRetriesExceededError{
+			err:   err,
+			tries: c.maxRetries,
+		}
+	}
+	retry := 0
+	if resp != nil {
+		retryAfter := resp.Header.Get("Retry-After")
+		retry, _ = strconv.Atoi(retryAfter) // we ignore the error here because an error already means retry is 0.
+	}
+	sleepFor := time.Duration(retry) * time.Second
+	if retry == 0 { // if we didn't get a Retry-After or it is zero, instead switch to exponential backoff
+		sleepFor = time.Duration(rand.Int63n(((1 << tries) - 1) * 10 * int64(time.Microsecond)))
+	}
+	if sleepFor > defaultMaxWait {
+		sleepFor = defaultMaxWait
+	}
+	time.Sleep(sleepFor)
+	atomic.AddUint64(triesPtr, 1)
+	return nil
+}
+
+func (c *Client) makeWriteRequest(bucket, org string, body io.Reader) (*http.Request, error) {
+	var err error
+	if c.contentEncoding == "gzip" {
+		body, err = gzip.CompressWithGzip(body, c.compressionLevel)
+		if err != nil {
+			return nil, err
+		}
+	}
+	u, err := makeWriteURL(c.url, bucket, org)
+	if err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequest("POST", u, body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "text/plain; charset=utf-8")
+
+	if c.contentEncoding == "gzip" {
+		req.Header.Set("Content-Encoding", "gzip")
+	}
+
+	req.Header.Set("User-Agent", c.userAgent)
+	req.Header.Set("Authorization", c.authorization)
+
+	return req, nil
+}
+
+func makeWriteURL(loc *url.URL, bucket, org string) (string, error) {
+	if loc == nil {
+		return "", errors.New("nil url")
+	}
+	u, err := url.Parse(loc.String())
+	if err != nil {
+		return "", err
+	}
+	params := url.Values{}
+	params.Set("bucket", bucket)
+	params.Set("org", org)
+
+	switch loc.Scheme {
+	case "http", "https":
+		u.Path = path.Join(u.Path, "/write")
+	case "unix":
+	default:
+		return "", fmt.Errorf("unsupported scheme: %q", u.Scheme)
+	}
+	u.RawQuery = params.Encode()
+	return u.String(), nil
+}
+
+// Close closes any idle connections on the Client.
+func (c *Client) Close() error {
```

Comment:
Don't think this is actually used anywhere?

Reply:
It is used.